contractcourt/channel_arbitrator.go

```diff
@@ -1269,15 +1321,171 @@ func (c *ChannelArbitrator) isPreimageAvailable(hash lntypes.Hash) (bool,
+// to ensure we're able to satisfy the HTLC timeout constraints for incoming vs
+// outgoing HTLCs.
+func (c *ChannelArbitrator) checkLocalChainActions(
+	height uint32, trigger transitionTrigger,
+	activeHTLCs map[HtlcSetKey]htlcSet,
+	commitsConfirmed bool) (ChainActionMap, error) {
+
+	// First, we'll check our local chain actions as normal. This will only
+	// examine HTLCs on our local commitment (timeout or settle).
+	localCommitActions, err := c.checkCommitChainActions(
+		height, trigger, activeHTLCs[LocalHtlcSet],
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Next, we'll examine the remote commitment (and maybe a dangling one)
+	// to see if the set difference of our HTLCs is non-empty. If so, then
+	// we may need to cnacel back some HTLCs if we decide go to chain.
+	remoteDanglingActions := c.checkRemoteDanglingActions(
+		height, activeHTLCs, commitsConfirmed,
+	)
+
+	// Finally, we'll merge the two set of chain actions.
+	localActions := make(ChainActionMap)
+	for chainAction, htlcs := range localCommitActions {
+		localActions[chainAction] = append(
+			localActions[chainAction], htlcs...,
+		)
+	}
+	for chainAction, htlcs := range remoteDanglingActions {
+		localActions[chainAction] = append(
+			localActions[chainAction], htlcs...,
+		)
+	}
+
+	return localActions, nil
+}
+
+// checkRemoteChainActions examines the set of remote commitments for any HTLCs
+// that are close to timing out. If we find any, then we'll return a set of
+// chain actions for HTLCs that are on our commitment, but not theirs to cancel
+// immediately.
+func (c *ChannelArbitrator) checkRemoteDanglingActions(
+	height uint32, activeHTLCs map[HtlcSetKey]htlcSet,
+	commitsConfirmed bool) ChainActionMap {
+
+	var (
+		pendingRemoteHTLCs []channeldb.HTLC
+		localHTLCs         = make(map[uint64]channeldb.HTLC)
+		remoteHTLCs        = make(map[uint64]channeldb.HTLC)
+		actionMap          = make(ChainActionMap)
+	)
+
+	// First, we'll construct to sets of the outgoing HTLCs on our local
+	// commitment, and the set of HTLCs on remote commitment(s).
+	for htlcSetKey, htlcs := range activeHTLCs {
+		if htlcSetKey.IsRemote {
+			continue
+		}
+
+		for _, htlc := range htlcs.outgoingHTLCs {
+			localHTLCs[htlc.HtlcIndex] = htlc
+		}
+	}
+	for htlcSetKey, htlcs := range activeHTLCs {
+		if !htlcSetKey.IsRemote {
+			continue
+		}
+
+		for _, htlc := range htlcs.outgoingHTLCs {
+			remoteHTLCs[htlc.HtlcIndex] = htlc
+		}
+	}
+
+	// With both sets constructed, we'll now compute the set difference of
+	// our two sets of HTLCs. This'll give us the HTLCs that exist on the
+	// remote commitment transaction, but not on ours.
+	for htlcIndex, htlc := range remoteHTLCs {
+		if _, ok := localHTLCs[htlcIndex]; ok {
+			continue
+		}
+
+		pendingRemoteHTLCs = append(pendingRemoteHTLCs, htlc)
+	}
+
+	// Finally, we'll examine al the pending remote HTLCs for those that
+	// are close to expiring. If we find any, then we'll recommend that
```

Comment:
should this read "have expired" instead of "close to expiring", since our outgoing broadcast delta is now zero?